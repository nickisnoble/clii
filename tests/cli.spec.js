const path = require("path");
const exec = require("child_process").exec;

describe("examples - dotenv", () => {
  it(".env file works", async () => {
    let result = await cli(["-f", resolveRootDir("examples/dotenv.mjs")]);
    expect(result.stdout).toMatchSnapshot();
  });
});

describe("examples - functions", () => {
  const prefixs = ["-f", resolveRootDir("examples/functions.mjs")];
  it("render cli", async () => {
    let result = await cli([...prefixs, "-h"]);
    expect(result.stdout).toMatchSnapshot();
  });
  it("task4 works", async () => {
    let result = await cli([...prefixs, "task4", "a", "b", "c"]);
    expect(result.stdout).toMatchSnapshot();
  });
  it("task5 works", async () => {
    let result = await cli([
      ...prefixs,
      "task5",
      "--str",
      "abc",
      "--num",
      "3",
      "--bool",
      "--arr",
      "x",
      "--arr",
      "y",
    ]);
    expect(result.stdout).toMatchSnapshot();
  });
  it("task6 works", async () => {
    let result = await cli([
      ...prefixs,
      "task6",
      "--foo",
      "abc",
      "--bar",
      "123",
      "ooo",
    ]);
    expect(result.stdout).toMatchSnapshot();
  });
});

describe("examples - options", () => {
  const prefixs = ["-f", resolveRootDir("examples/options.mjs")];
  it("render cli", async () => {
    let result = await cli([...prefixs, "-h"]);
    expect(result.stdout).toMatchSnapshot();
  });
  it("it works", async () => {
    let result = await cli([
      ...prefixs,
      "--str",
      "abc",
      "--num",
      "3",
      "--bool",
      "--arr",
      "x",
      "--arr",
      "y",
    ]);
    expect(result.stdout).toMatchSnapshot();
  });
});

describe("examples - globals", () => {
  const prefixs = ["-f", resolveRootDir("examples/globals.mjs")];
  it("render cli", async () => {
    let result = await cli([...prefixs, "-h"]);
    expect(result.stdout).toMatchSnapshot();
  });
  it("it works", async () => {
    let result = await cli(prefixs);
    expect(result.stdout).toMatchSnapshot();
  });
});

function resolveRootDir(file) {
  return path.resolve(__dirname, "..", file);
}

function cli(args, cwd) {
  return new Promise((resolve) => {
    exec(
      `node ${resolveRootDir("dist/main.js")} ${args.join(" ")}`,
      { cwd },
      (error, stdout, stderr) => {
        resolve({
          code: error && error.code ? error.code : 0,
          error,
          stdout,
          stderr,
        });
      }
    );
  });
}
