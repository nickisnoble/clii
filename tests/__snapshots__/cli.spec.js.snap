// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`examples - functions render cli 1`] = `
"Usage: main.js <cmd> [options]

Commands:
  main.js task1 <foo>            One parameter
  main.js task2 [foo]            Parameter with default value
  main.js task3 <arr...>         Array parameter
  main.js task4 <foo> <bars...>  Multiple parameters
  main.js task5 [options]        Command with options
  main.js task6 <pos> [options]  Both parameters and options

Options:
      --version  Show version number                                   [boolean]
  -f, --file     Specific cmru file                                     [string]
  -w, --workdir  Specific working directory                             [string]
      --verbose  Echo command                                          [boolean]
      --silent   Suppress all normal output                            [boolean]
  -h, --help     Show help                                             [boolean]
"
`;

exports[`examples - functions task4 works 1`] = `
"a b,c
"
`;

exports[`examples - functions task4 works2 1`] = `
"-a -b -c -x -y -z
"
`;

exports[`examples - functions task5 works 1`] = `
"{\\"str\\":\\"abc\\",\\"num\\":3,\\"bool\\":true,\\"arr\\":[\\"x\\",\\"y\\"]}
"
`;

exports[`examples - functions task5 works2 1`] = `
"{\\"str\\":\\"-a -b -c\\"}
"
`;

exports[`examples - functions task6 works 1`] = `
"{\\"options\\":{\\"foo\\":\\"abc\\",\\"bar\\":123},\\"pos\\":\\"ooo\\"}
"
`;

exports[`examples - settings it works 1`] = `
"{ str: 'abc', bool: true, num: 3, arr: [ 'x', 'y' ] }
"
`;

exports[`examples - settings render cli 1`] = `
"Usage: main.js <cmd> [options]

Options:
      --version  Show version number                                   [boolean]
  -f, --file     Specific cmru file                                     [string]
  -w, --workdir  Specific working directory                             [string]
      --verbose  Echo command                                          [boolean]
      --silent   Suppress all normal output                            [boolean]
      --str      String variable                     [string] [default: \\"0.1.0\\"]
      --bool     Boolean variable                     [boolean] [default: false]
      --num      Number variable                           [number] [default: 3]
      --arr      Array varialbe                            [array] [default: []]
  -h, --help     Show help                                             [boolean]
"
`;
